### PR TITLE
Second Phase of RLC inference

### PR DESCRIPTION
**Pull request by Nargeshdb**
This PR infers:

- `@NotOwning` for return types
- `@MustCallAlias `for both return types and parameter types
- `@Owning` for constructor parameters when there are multiple owning fields.

### CODE REVIEW

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (kelloggm) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:145]: Clarify whether these are must or may aliases (I think must?).

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (kelloggm) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:264]: Add a clause to this sentence saying why (so that `@MustCallAlias` annotations can be inferred later, I think)

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (kelloggm) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:398]: Rather than changing this from 3 to 4, maybe change this to be non-specific about the size of the list, so that we don't have to update it in the future: `performs the following computations:`

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (kelloggm) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:448]: I'm concerned about this early return, because it means that the "otherwise" part of the javadoc doesn't actually happen if `rhsObligation` isn't associated with a parameter. I'm struggling to articulate a situation in which that's incorrect, but it doesn't seem to preserve the same semantics here.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (kelloggm) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:475]: typo: "alias of" -> "aliased to"

Comment (kelloggm) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:475]: also, probably a good idea to say "must-aliased" rather than just "aliased"

**Review by kelloggm — CHANGES_REQUESTED** [review]

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (Nargeshdb) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:448]: I think it's fine because we are only tracking formal parameter aliases, and if the rhsObligation is not an alias of any of the formal parameters, it won't exist in the obligations set. So skipping `mcca.removeObligationsContainingVar(obligations, (LocalVariableNode) rhs);` line doesn't break anything.

Comment (kelloggm) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:449]: Can you add a comment just above the `return;` with this information?

**Review by kelloggm — APPROVED** [review]
LGTM. @msridhar should also review before merging.

**Comment by mernst** [comment]
I should review after Manu does.

File: checker/tests/ainfer-resourceleak/non-annotated/AddNotOwning.java
Comment (msridhar) [inline checker/tests/ainfer-resourceleak/non-annotated/AddNotOwning.java:1]: For all these tests in the `non-annotated` folder it's unclear to me what is being tested.  Can you explain how these tests work, or is it documented somewhere?

**Review by msridhar — COMMENTED** [review]
Mostly looks good!  Just a couple of comments

File: checker/tests/ainfer-resourceleak/non-annotated/AddNotOwning.java
Comment (kelloggm) [reply checker/tests/ainfer-resourceleak/non-annotated/AddNotOwning.java:1]: It's documented here: https://github.com/typetools/checker-framework/blob/master/checker/tests/ainfer-README

That document describes the "TestChecker", but for these tests the same process is used with the RLC instead. This README is also symlinked from the ainfer-resourceleak/ directory.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (Nargeshdb) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:279]: Based on the algorithm for adding the `@NotOwning` annotation, we don't need to ensure it's an owning field, just a field with non-empty must-call type should be enough. The documentation is not correct, I updated it.

File: checker/tests/ainfer-resourceleak/non-annotated/MustCallAliasOnRegularExits.java
Comment (msridhar) [inline checker/tests/ainfer-resourceleak/non-annotated/MustCallAliasOnRegularExits.java:18]: Why do we have the comment if we are already writing `@Owning` in the test case?

**Review by msridhar — COMMENTED** [review]
Couple questions on the tests

File: checker/tests/ainfer-resourceleak/non-annotated/MustCallAliasOnReceiver.java
Comment (Nargeshdb) [reply checker/tests/ainfer-resourceleak/non-annotated/MustCallAliasOnReceiver.java:7]: Right now we don't have support adding `@MustCallAlias` annotation to the receiver parameter, so I removed the related test from this file to avoid crashes, I added it back, and instead I added `// @skip-test` in this file.

File: checker/tests/ainfer-resourceleak/non-annotated/MustCallAliasOnReceiver.java
Comment (Nargeshdb) [reply checker/tests/ainfer-resourceleak/non-annotated/MustCallAliasOnReceiver.java]: This comment added by mistake and maneFactory doesn't throw an exception here.

File: checker/tests/ainfer-resourceleak/non-annotated/MustCallAliasOnRegularExits.java
Comment (Nargeshdb) [reply checker/tests/ainfer-resourceleak/non-annotated/MustCallAliasOnRegularExits.java]: This comment added by mistake.

**Review by msridhar — CHANGES_REQUESTED** [review]
Let's please add a comment at the top of each test source file indicating what is being tested by that file.  Otherwise, this PR is good to go from my perspective

**Comment by msridhar** [comment]
@mernst I think you can start reviewing this one.  I just have one outstanding comment on the test files that should be straightforward to address

**Review by msridhar — APPROVED** [review]

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (mernst) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:156]: Should this be named "returnObligationToParameter"?

More generally, please discuss the choice to use obligations rather than nodes as the keys.  The documentation says that the map is from return nodes to method parameters, but the implementation is different.  Why are obligations needed?  Is it because they indicate resource aliasing relationships?  Or is it for some other reason?

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (mernst) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:570]: Why is the name "addOrUpdate"?  The first sentence of the documentation lacks "update", and I don't see that it adds much (if anything) to the clarity at call sites.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (mernst) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:419]: Does this need to be re-done if some other part of the inference discovers an owning field?

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (mernst) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:513]: Writing a *method declaration* here is inconsistent with the documentation that says the annotation is added to the *return type*.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (mernst) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:538]: The value of `mustCallValues` was known when constructing `disposedFields`.  Would it make sense to record it it `dispasedFields` so it does not need to be recomputed here?

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (mernst) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:617]: I am confused by this comment.  Should it be, "Computes ownership transfer at the method call to infer @Owning annotations for formal parameters of the current method, if the parameter is passed into the call and the corresponding formal parameter of the callee is @Owning."?

Comment (Nargeshdb) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java]: It is correct, I updated the comment to what you suggested.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (Nargeshdb) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:492]: It's because we need to iterate over all the aliases of the given obligation.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (Nargeshdb) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java]: It won't be re-done in this iteration, in the next iteration it will infer an @Owning annotation if the number of owning fields is greater than 1. I added a new test file(`ReplaceMustCallAliasAnnotation.java`) to check this scenario.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (Nargeshdb) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:160]: They are needed to check aliasing relationships of return nodes with parameters.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (Nargeshdb) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:107]: It's pointing to the fields, I removed "at the entry ..." to prevent future confusion.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (Nargeshdb) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:114]: Since we don't iterate over this field in our computation, I think using a HashSet should fine.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (Nargeshdb) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java:101]: This might cause nondeterministic ordering of fields in `@EnsuresCalledMethods` annotations, and to prevent this, I've switched to using `LinkedHashSet`.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java
Comment (Nargeshdb) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java]: I changed the `deposedFields` set to a map to keep the `must-call` value and save some computation here. Thanks for the suggestion.

**Review by mernst — DISMISSED** [review]

**Review by mernst — DISMISSED** [review]

**Comment by msridhar** [comment]
@mernst can I get one more approval from you?  I had to remove some logic related to resource variables that should be unnecessary now that #6273 has landed

**Review by mernst — APPROVED** [review]
Thanks for the simplifictation!

**Pull request merged**
